diff --git a/config.h b/config.h
index 625b1a4..0a61072 100644
--- a/config.h
+++ b/config.h
@@ -38,7 +38,7 @@ static const unsigned int alphas[][2]      = {
 };
 /* -l option; if nonzero, dmenu uses vertical list with given number of lines */
 static unsigned int lines      = 0;
-static unsigned int lineheight = 0;         /* -h option; minimum height of a menu line     */
+static unsigned int lineheight = 22;         /* -h option; minimum height of a menu line     */
 
 /*
  * Characters not considered part of a word while deleting words
